---
title: "Soil GC Fractions Community Diversity"
author: "Dylan Bodington"
date: "`r Sys.Date()`"
out:
  html_document:
    toc: true
    toc_depth: 2
    code_folding: hide
    fig_caption: true
    
citation_package: bibtex
bibliography: bib.bib
link-citations: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(devtools)
library(ggplot2)
library(ggpubr)
library(ggsci)
library(RColorBrewer)
library(microshades)
library(tidyr)
library(purrr)
library(stringr)
library(dplyr)
library(phyloseq)
library(microbiome)
library(speedyseq)
library(microViz)
library(metagMisc)
library(Biostrings)
library(phyloseq.mate)
```

## Introduction

After cleaning and filtering using trimgalore, the 16S were processed using the standard [dada2 pipeline](https://benjjneb.github.io/dada2/tutorial.html) to create phyloseq objects of all ASVs. When rdp cannot reliably assign taxonomy (in this case down to Genus), levels below the lowest reliable assignment are designated NA.
[Download link for sequence files and metadata.](01_data/01_input)

```{r ps_objects, message=FALSE, warning=FALSE, include=FALSE}
ps_raw <- readRDS('./01_data/phyloseq_gtdb.Rds')
set.seed(90210)
ps_16S <- ps_raw %>%
  rarefy_even_depth(sample.size = min(sample_sums(ps_raw))) %>%
  tax_filter(min_sample_abundance = 2)

sort_taxa <- read.csv("00_ref/GTDB_bac120_arc53_ssu_r207_clade_sort.csv", header = TRUE)
```

```{r import_mag,  message=FALSE, warning=FALSE, include=FALSE}
tax_merge <- merge_phyloseq(tax_table(ps_16S),tax_table(as.matrix(read.table(file = "01_data/gc_mag_tax.csv", header = TRUE, sep = ",", row.names = 1))))
sample_data_merge <- sample_data(read.table(file = "00_ref/phyloseq_metadata.csv", header = TRUE, sep = ",", row.names = 1))

otu_mag <- otu_table(round(as.matrix(read.table(file = "01_data/gc_mag_otu.csv", header = TRUE, sep = ",", row.names = 1))), taxa_are_rows = FALSE)
otu_mag_merge_ps_16S <- otu_table(matrix(0, length(rownames(otu_mag)), ncol(otu_table(ps_16S)), dimname = list(rownames(otu_mag),colnames(otu_table(ps_16S)))), taxa_are_rows = FALSE)
otu_ps_16S_merge_mag <- otu_table(matrix(0, length(rownames(otu_table(ps_16S))), ncol(otu_mag), dimname = list(rownames(otu_table(ps_16S)),colnames(otu_mag))), taxa_are_rows = FALSE)
otu_merge <- merge_phyloseq(merge_phyloseq(otu_table(ps_16S), otu_mag_merge_ps_16S), merge_phyloseq(otu_ps_16S_merge_mag, otu_mag))

ps_merge <- phyloseq(otu_merge, sample_data_merge, tax_merge)

ps_set <- subset_samples(ps_merge, sample %in% c("4.5_M5", "4.5_M11", "7.5_M4", "7.5_M14", "CM_M5", "CM_M11", "LD_M4", "LD_M13"))

sample_names(ps_set) <- c("4.5_1", "4.5_20", "7.5_1", "7.5_20", "CM_1", "CM_20", "LD_1", "LD_20")
sample_data(ps_set)$sample <- sample_names(ps_set)
sample_data(ps_set)$name <- c("GcA1", "GcA20", "GcB1", "GcB20", "GcC1", "GcC20", "GcL1", "GcL20")
sample_data(ps_set)$fraction <- c("1", "20", "1", "20", "1", "20", "1", "20")
sample_data(ps_set)$label <- c("Fill1", "Fill2", "Fill3", "Fill4", "Fill5", "Fill6", "Fill7", "Fill8")

ps_merge <- merge_phyloseq(ps_merge, ps_set)

sample_data(ps_merge) <- sample_data(read.table(file = "00_ref/phyloseq_metadata.csv", header = TRUE, sep = ",", row.names = 1))

sample_data(ps_merge)$name <- factor(sample_data(ps_merge)$name, levels = str_sort(sample_data(ps_merge)$name, numeric = TRUE))
ps_merge <- ps_arrange(ps_merge, name)
sample_data(ps_merge)$soil <- factor(sample_data(ps_merge)$soil, levels = c("GcA (Craibstone pH 4.5)", "GcB (Craibstone pH 7.5)", "GcC (Corsemaul)", "GcL (Lower Dell)"))
sample_data(ps_merge)$label <- factor(sample_data(ps_merge)$label, levels = sample_data(ps_merge)$label)

ps_all <- subset_samples(ps_raw, !sample %in% c("4.5_0.1", "4.5_0.2", "4.5_0.3", "7.5_0.1", "7.5_0.2", "7.5_0.3", "CM_0.1", "CM_0.2", "CM_0.3", "LD_0.1", "LD_0.2", "LD_0.3"))
ps_merge <- subset_samples(ps_merge, !sample %in% c("4.5_0.1", "4.5_0.2", "4.5_0.3", "7.5_0.1", "7.5_0.2", "7.5_0.3", "CM_0.1", "CM_0.2", "CM_0.3", "LD_0.1", "LD_0.2", "LD_0.3"))
#ps_merge <- subset_samples(ps_merge, !label %in% c("GcArep1", "GcArep2", "GcArep3", "GcBrep1", "GcBrep2", "GcBrep3", "GcCrep1", "GcCrep2", "GcCrep3", "GcLrep1", "GcLrep2", "GcLrep3"))
save(ps_all, ps_merge, file = "./01_data/phyloseq_all.Rdata")
```

## Alpha Diversity
```{r rarefaction_curve, fig.height=8, fig.width=12}
otu <- t(abundances(ps_all))
rare_curve <- vegan::rarecurve(otu, step = 50)
rare <- lapply(rare_curve, function(x){
  b <- as.data.frame(x)
  b <- data.frame(OTU = b[,1], raw.read = rownames(b))
  b$raw.read <- as.numeric(gsub("N", "",  b$raw.read))
  return(b)
  }
)
names(rare) <- rownames(otu)

rare <- map_dfr(
          rare,
          function(x){
            z <- data.frame(x)
            return(z)
            },
          .id = "sample"
          )
rare$soil = str_extract(rare$sample, pattern = "^[^_]+")

rare_labels <- as_labeller(c(`4.5` = "GcA", `7.5` = "GcB", `CM` = "GcC", `LD` = "GcL"))

rare_data <- rare %>%
            group_by(sample) %>%
            arrange(desc(raw.read)) %>%
            filter(row_number()==1)

rare_plot <- ggplot(rare) +
        geom_line(
          aes(
            x = raw.read,
            y = OTU,
            colour = sample
          )
        ) +
        geom_text(
          data = rare_data,
          aes(
            x = raw.read + 6000, y = OTU, label = rare_data$sample
          )
        ) +
        facet_wrap(
          vars(soil),
          nrow = 1,
          labeller = rare_labels
        )
```


```{r alpha_16S, fig.height=8, fig.width=12}
alpha_all <- alpha(ps_all, index = "all")
rownames(alpha_all) <- sample_data(ps_all)$name
alpha_all$source <- sample_data(ps_all)$soil
alpha_all$name <- sample_data(ps_all)$label
alpha_all <- alpha_all %>%
  relocate(source, name, .before = observed)

write.table(alpha_all,
            file = "./02_out/alpha_diversity.csv",
            sep = ",",
            quote = T,
            row.names = T,
            col.names = T
            )
```

```{r}
ordcap = ordinate(ps_filter(ps_16S, GC != "NA"), "CAP", "bray", ~GC)
bray_plot <- plot_ordination(ps_16S, ordcap, "samples", color="soil")

ggplot2::ggsave(bray_plot, file = "./02_out/03_bray.png", height = 8, width = 8)
```

```{r shannon, fig.height=4, fig.width=12}
shannon_plot <- plot_richness(ps_merge,
                                  x = 'label',
                                  measures=c("Shannon"),
                                  title = "Shannon Diversity of Soil Community Fractions") +
  scale_x_discrete(labels = c(sample_data(ps_merge)$label)) +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x") +
  scale_color_npg() +
  guides(colour = "none") +
  labs(x = "",
       y = "Shannon Diversity") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -90))
```


```{r observed, fig.height=4, fig.width=12}
observed_plot <- plot_richness(ps_merge,
                                  x = 'label',
                                  measures=c("Observed"),
                                  title = "Observed Richness of Soil Community Fractions") +
  scale_x_discrete(labels = c(sample_data(ps_merge)$label)) +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x") +
  scale_color_npg() +
  guides(colour = "none") +
  labs(x = "",
       y = "Richness") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -90))
```

## Community Composition

```{r gtdb_relabund}
#phylum_levels <- append(sort_taxa$Phylum[nzchar(sort_taxa$Phylum)], c("Unassigned", "Other"))



write.table(ps_all %>%
              tax_fix() %>%
              tax_transform(trans = "compositional", rank = "Phylum") %>%
              psmelt() %>%
              select(name, Phylum, Abundance) %>%
#              arrange(relabund_table, match(relabund_table$Phylum, phylum_levels)) %>%
              pivot_wider(names_from = Phylum, values_from = Abundance),
            file = "./02_out/ps_16S.relative_abundance.phylum.csv",
            sep = ",",
            quote = F,
            row.names = F,
            col.names = T
            )
```

```{r phylum_relabund, fig.height=6, fig.width=20}
ps <- ps_merge %>%
  mutate_tax_table(across(everything(), \(x) str_replace(x,'_[A-Z]', '')))
  
phylum_levels <- append(levels(factor(tax_table(ps)[,"Phylum"])), c("Other"))

ps <- ps %>%
  tax_fix() %>%
  tax_transform(trans = "compositional") %>%
  transform_sample_counts(function(x) x * 100) %>%
  tax_agg(rank = "Phylum", top_N = 24)

my_pal <- c(pal_simpsons("springfield")(16)[-5], "steelblue2", pal_flatui("default")(8), "dimgray")

taxa_plot <- plot_bar(ps,
                          x = "label", 
                          fill="top") +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x") +
  labs(fill = "Taxonomy (Phylum)    ",
       y = "Relative Abundance (%)") +
  theme() +
  scale_fill_manual(values = my_pal) +
  guides(fill = guide_legend(ncol = 1, byrow = TRUE))

taxa_plot$data$top[taxa_plot$data$top %in% c("other", NA)] <- rep("Other", length(taxa_plot$data$top[taxa_plot$data$top %in% c("other", NA)]))
taxa_plot$data$top <- str_pad(taxa_plot$data$top, 21, "right")
phylum_levels <- str_pad(phylum_levels, 21, "right")
taxa_plot$data$top <- factor(taxa_plot$data$top, levels = phylum_levels)
taxa_plot$layers[[1]]$geom_params$width <- 1
```

```{r domain_relabund, fig.height=6, fig.width=20}
ps <- ps_merge %>%
  mutate_tax_table(across(everything(), \(x) str_replace(x,'_[A-Z]', '')))
  
domain_levels <- append(levels(factor(tax_table(ps)[,"Kingdom"])), c("Other"))

ps <- ps %>%
  tax_fix() %>%
  tax_transform(trans = "compositional") %>%
  transform_sample_counts(function(x) x * 100) %>%
  tax_agg(rank = "Kingdom", top_N = 2)

my_pal <- c(pal_simpsons("springfield")(2), "dimgray")

taxa_plot_domain <- plot_bar(ps,
                          x = "label", 
                          fill="top") +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x") +
  labs(fill = "Taxonomy (Domain)    ",
       y = "Relative Abundance (%)") +
  theme() +
  scale_fill_manual(values = my_pal) +
  guides(fill = guide_legend(ncol = 1, byrow = TRUE))

taxa_plot_domain$data$top[taxa_plot_domain$data$top %in% c("other", NA)] <- rep("Other", length(taxa_plot_domain$data$top[taxa_plot_domain$data$top %in% c("other", NA)]))
taxa_plot_domain$data$top <- str_pad(taxa_plot_domain$data$top, 21, "right")
domain_levels <- str_pad(domain_levels, 21, "right")
taxa_plot_domain$data$top <- factor(taxa_plot_domain$data$top, levels = domain_levels)
taxa_plot_domain$layers[[1]]$geom_params$width <- 1
```

Filler

## Plots for publication

Download links below plot

### Fig 1

```{r pub_fig_1, fig.height=6, fig.width=13}
zp1 <- shannon_plot +
  scale_y_continuous(limits = c(4.2, 6.8),
                     breaks = c(5, 6)) +
  labs(title = "Catchy Graph Title A",
       y = "Shannon\ndiversity") +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_text(size = 12),
    plot.margin = unit(c(1,1,-0.1,1), "lines"))

zp2 <- observed_plot +
  scale_y_continuous(limits = c(100, 1200),
                     breaks = c(500, 1000)) +
  labs(y = "Richness") +
  theme(
    title = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_blank(),
    plot.margin = unit(c(1,1,-0.1,1), "lines"))

zp3 <- taxa_plot +
  guides(fill = guide_legend(nrow = 3, byrow = TRUE)) +
  labs(x = "",
       y = "Relative Abundance (%)",
       legend = "Phylum") +
  theme(strip.text.x = element_blank(),
        axis.text.x = element_text(
          colour = "black",
          vjust = 0.5
          ),
        legend.position = "bottom",
        legend.justification = "right",
        legend.spacing.x = unit(0.5, 'cm'),
        legend.text = element_text(
          size = 9,
          margin = margin(l = 0.1, unit ="cm")
          ),
        plot.margin = unit(c(-0.1,1,1,1), "lines"))

zp <- ggarrange(zp1, zp2, zp3,
          nrow = 3,
          heights = c(0.32, 0.22, 1),
          align = "v")

ggplot2::ggsave(zp, file = "./02_out/01_alpha_diversity.png", height = 8, width = 14)
```

[Download this plot.](https://github.com/bodington/gc_16S/raw/main/02_out/01_alpha_diversity.png)

```{r pub_fig_2, fig.height=6, fig.width=13}
zp <- rare_plot +
  labs(title = "Catchy Rare Graph Title",
        x = "Sample Size",
        y = "Observed OTUs") +
  scale_x_continuous(limits = c(0, 75000)) +
  theme(strip.text.x = element_text(size = 12),
        legend.position = "none",
        plot.margin = unit(c(1,1,1,1), "lines"))
ggplot2::ggsave(zp, file = "./02_out/02_rare.png", height = 7, width = 14)
```

[Download this plot.](https://github.com/bodington/gc_16S/raw/main/02_out/02_rare.png)

```{r pub_fig_s1, fig.height=5, fig.width=13}
zp1 <- shannon_plot +
  scale_y_continuous(limits = c(4.2, 6.8),
                     breaks = c(5, 6)) +
  labs(title = "Catchy Graph Title A",
       y = "Shannon\ndiversity") +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_text(size = 12),
    plot.margin = unit(c(1,1,-0.1,1), "lines"))

zp2 <- observed_plot +
  scale_y_continuous(limits = c(100, 1200),
                     breaks = c(500, 1000)) +
  labs(y = "Richness") +
  theme(
    title = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_blank(),
    plot.margin = unit(c(1,1,-0.1,1), "lines"))

zp3 <- taxa_plot_domain +
  guides(fill = guide_legend(nrow = 1, byrow = TRUE)) +
  labs(x = "",
       y = "Relative Abundance (%)",
       legend = "Phylum") +
  theme(strip.text.x = element_blank(),
        axis.text.x = element_text(
          colour = "black",
          vjust = 0.5
          ),
        legend.position = "bottom",
        legend.justification = "right",
        legend.spacing.x = unit(0.5, 'cm'),
        legend.text = element_text(
          size = 9,
          margin = margin(l = 0.1, unit ="cm")
          ),
        plot.margin = unit(c(-0.1,1,1,1), "lines"))

zp <- ggarrange(zp1, zp2, zp3,
          nrow = 3,
          heights = c(0.32, 0.22, 0.6),
          align = "v")

ggplot2::ggsave(zp, file = "./02_out/s1_alpha_diversity.png", height = 8, width = 14)
```

```{r, message=FALSE, warning=FALSE, include=FALSE}
ps <- ps_merge %>%
  transform_sample_counts(function(x) ((x * 100) / sum(x)))
my_subset <- subset(t(otu_table(ps)), rownames(t(otu_table(ps))) %in% mag_asv)
ps_mag <- merge_phyloseq(my_subset, tax_table(ps), sample_data(ps), refseq(ps_16S))
```



