---
title: "Soil GC Fractions Community Diversity"
author: "Dylan Bodington"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 2
    code_folding: hide
    fig_caption: true
    
citation_package: bibtex
bibliography: bib.bib
link-citations: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2); theme_set(theme_light())
library(ggpubr)
library(tidyr)
library(stringr)
library(dplyr)
library(phyloseq)
library(speedyseq)
#library(phyloniche)
library(metagMisc)
#library(devtools)
```

## Introduction

After cleaning and filtering using trimgalore, the 16S were processed using the standard [dada2 pipeline](https://benjjneb.github.io/dada2/tutorial.html) to create phyloseq objects of all ASVs. When rdp cannot reliably assign taxonomy (in this case down to Genus), levels below the lowest reliable assignment are designated NA. [This script](https://github.com/bodington/sponge/blob/main/R/phyloseq_taxonomy_imputation.R) modifies the tax table of a phyloseq object to replace NA with the lowest assigned level + "_unidentified" eg "Bacteria, Cyanobacteria, Cyanobacteria_unidentified"


```{r ps_objects, warning=FALSE}
ps_16S <- readRDS('./02_output/phyloseq.Rds')

source("./R/phyloseq_taxonomy_imputation.R")  #Replacing NA in taxonomy assignment with the lowest level of assignment

ps_16S <- phyloseq_taxonomy_imputation(ps_16S)

#Merging all samples of each host
#ps_16S_merged <- merge_samples(ps_16S, "host")


# Removing low abundance (<0.001% rel abund) ASVs. Uses functions from metagMisc
ps_16S_abund <- phyloseq_filter_taxa_rel_abund(ps_16S, frac = 0.00001)
```

## Alpha Diversity

The [Shannon Diversity Index](https://en.wikipedia.org/wiki/Diversity_index) of the communities in each sponge host were estimated using the 16S gene. The host communities were ordered from low to high diversity, and this order and colour scheme was used to maintain consistency for later analyses. The shape indicates the [microbial abundance](https://pubmed.ncbi.nlm.nih.gov/15825640/) of the host .

```{r shannon_16S, fig.height=5, fig.width=10}
shannon_16S_plot <- plot_richness(ps_16S, x = 'name', color = 'soil', measures=c("Shannon"), title = "", sortby = ) +
  geom_point(
    aes(
      colour = soil
    ), 
    size = 3
  ) +
#  scale_colour_gradient2() +
  guides(
    colour = "none"
  ) +
  labs(
    title = "Diversity of Whole Microbiome (16S)",
    shape = "Microbial Abundance",
    x = "Sample",
    y = "Shannon Diversity Index"
  ) +
  theme(
    axis.text.x = element_text(angle = -30,
                  ),
    strip.text = element_blank(
    )
  )
  #ylim(0, 6.2)
shannon_16S_plot$data$name <- factor(shannon_16S_plot$data$name, levels = str_sort(sample_data(ps_16S)$name, numeric = TRUE))
shannon_16S_plot
```

## Community Composition

When displaying the taxonomic makeup of the communities at phylum level, I split Pseudomonadota into classes. This should be done on a copy of the phyloseq object, which was already created in the previous step.

```{r proteobacteria_phylum, warning=FALSE}
ps_16S_proteo <- ps_16S

plot.tax = as.data.frame(tax_table(ps_16S))
plot.tax = data.frame(lapply(plot.tax, as.character), stringsAsFactors = F)
plot.tax$Phylum[plot.tax$Phylum=="Proteobacteria"] = plot.tax$Class[plot.tax$Phylum=="Proteobacteria"] 
plot.tax[] = lapply(plot.tax, factor)
plot.tax.table = tax_table(plot.tax)
rownames(plot.tax.table) = rownames(tax_table(ps_16S))
colnames(plot.tax.table) = colnames(tax_table(ps_16S))

tax_table(ps_16S_proteo) = plot.tax.table
rm(list = c("plot.tax" , "plot.tax.table"))
```

```{r phylum_relabund, fig.height=8, fig.width=12}
ps <- tax_glom(ps_16S_proteo, taxrank = "Phylum", bad_empty = "Bacteria_unidentified")
#ps <- merge_samples2(ps, "host", funs = list())
ps <- transform_sample_counts(ps, function(x) x / sum(x))
ps <- phyloseq_filter_top_taxa(ps, n = 20)

write.table(ps %>%
              psmelt() %>%
              select(Phylum, Sample, Abundance) %>%
              spread(Sample, Abundance),
	file = "./02_output/ps_16S.relative_abundance.clade.csv", sep = ",", quote = F, row.names = F, col.names = T)

taxa_16S_plot <- plot_bar(ps, x = "name", fill="Phylum") +
  labs(
    title = "Community Composition (16S)",
    fill = "Phylum",
    x = "Sample",
    y = "Relative Abundance"
  ) +
  theme(
    axis.text.x = element_text(
      angle = -30,
      face = "italic"
    )
  ) +
 scale_fill_manual(
    values = c("#89C5DA", "#DA5724", "#74D944", "#CE50CA", "#3F4921", "#C0717C", "#7FDCC0", "#5F7FC7", "#673770", "#D3D93E", "#38333E", "#508578", "#D7C1B1", "#689030", "#AD6F3B", "#CD9BCD", "#D14285", "#6DDE88", "#652926", "#CBD588"))
taxa_16S_plot$data$name <- factor(shannon_16S_plot$data$name, levels = str_sort(sample_data(ps_16S)$name, numeric = TRUE))
taxa_16S_plot
```