---
title: "Soil GC Fractions Community Diversity"
author: "Dylan Bodington"
date: "`r Sys.Date()`"
out:
  html_document:
    toc: true
    toc_depth: 2
    code_folding: hide
    fig_caption: true
    
citation_package: bibtex
bibliography: bib.bib
link-citations: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2); theme_set(theme_light())
library(ggpubr)
library(ggsci)
library(RColorBrewer)
library(tidyr)
library(stringr)
library(dplyr)
library(phyloseq)
library(microbiome)
library(speedyseq)
library(microViz)
library(metagMisc)
library(Biostrings)
source("./R/phyloseq.mate.R")
```

## Introduction

After cleaning and filtering using trimgalore, the 16S were processed using the standard [dada2 pipeline](https://benjjneb.github.io/dada2/tutorial.html) to create phyloseq objects of all ASVs. When rdp cannot reliably assign taxonomy (in this case down to Genus), levels below the lowest reliable assignment are designated NA. [This script](https://github.com/bodington/gc_16S/blob/main/R/phyloseq_taxonomy_imputation.R) modifies the tax table of a phyloseq object to replace NA with the lowest assigned level + "_unidentified" eg "Bacteria, Cyanobacteria, Cyanobacteria_unidentified"


```{r ps_objects, message=FALSE, warning=FALSE, include=FALSE}
ps_16S <- readRDS('./01_data/phyloseq_gtdb.Rds')
```
```{r ps_filler_samples, message=FALSE, warning=FALSE, include=FALSE}
#ps_set <- subset_samples(ps_16S, label %in% c("Whole Replicate 2", "Whole Replicate 3"))
ps_set <- ps_16S
sample_names(ps_set) <- c("4.5_1", "4.5_20", "7.5_1", "7.5_20", "CM_1", "CM_20", "LD_1", "LD_20")
sample_data(ps_set)$sample <- sample_names(ps_set)
sample_data(ps_set)$name <- c("GcA1", "GcA20", "GcB1", "GcB20", "GcC1", "GcC20", "GcL1", "GcL20")
sample_data(ps_set)$fraction <- c("1", "20", "1", "20", "1", "20", "1", "20")
sample_data(ps_set)$label <- c("Fill1", "Fill2", "Fill3", "Fill4", "Fill5", "Fill6", "Fill7", "Fill8")

ps_16S <- merge_phyloseq(ps_16S, ps_set)
sample_data(ps_16S)$name <- factor(sample_data(ps_16S)$name, levels = str_sort(sample_data(ps_16S)$name, numeric = TRUE))
ps_16S <- ps_arrange(ps_16S, name) %>%
  tax_fix()
```


```{r import_mag,  message=FALSE, warning=FALSE, include=FALSE}
tax_merge <- merge_phyloseq(tax_table(ps_16S),tax_table(as.matrix(read.table(file = "01_data/gc_mag_tax.csv", header = TRUE, sep = ",", row.names = 1))))
sample_data_merge <- sample_data(read.table(file = "00_ref/phyloseq_metadata.csv", header = TRUE, sep = ",", row.names = 1))

otu_mag <- otu_table(round(as.matrix(read.table(file = "01_data/gc_mag_otu.csv", header = TRUE, sep = ",", row.names = 1))), taxa_are_rows = FALSE)
otu_mag_merge_ps_16S <- otu_table(matrix(0, length(rownames(otu_mag)), ncol(otu_table(ps_16S)), dimname = list(rownames(otu_mag),colnames(otu_table(ps_16S)))), taxa_are_rows = FALSE)
otu_ps_16S_merge_mag <- otu_table(matrix(0, length(rownames(otu_table(ps_16S))), ncol(otu_mag), dimname = list(rownames(otu_table(ps_16S)),colnames(otu_mag))), taxa_are_rows = FALSE)
otu_merge <- merge_phyloseq(merge_phyloseq(otu_table(ps_16S), otu_mag_merge_ps_16S), merge_phyloseq(otu_ps_16S_merge_mag, otu_mag))

ps_merge <- phyloseq(otu_merge, sample_data_merge, tax_merge) %>%
#  ps_filter(!(label %in% c("Mixed", "Whole Replicate 2", "Whole Replicate 3"))) %>%
  tax_fix()

labels <- append(sample_data(ps_merge)$label[!sample_data(ps_merge)$label %in% c("Whole", "Genome (MAG)")], "Whole", 0)

sample_data(ps_merge)$name <- factor(sample_data(ps_merge)$name, levels = str_sort(sample_data(ps_merge)$name, numeric = TRUE))
labels = c(c("Whole"), sample_data(ps_merge)$label[!sample_data(ps_merge)$label %in% c("Whole")])
sample_data(ps_merge)$soil <- factor(sample_data(ps_merge)$soil, levels = c("Craibstone pH 4.5", "Craibstone pH 7.5", "Corsemaul", "Lower Dell"))
sample_data(ps_merge)$label <- factor(sample_data(ps_merge)$label, levels = labels)
ps_merge <- ps_arrange(ps_merge, name)
```

## Alpha Diversity

```{r alpha_16S, fig.height=8, fig.width=12}
otu_tab <- t(abundances(ps_16S))
png("./02_out/rare.png")
p <- vegan::rarecurve(otu_tab, 
                      step = 50, label = FALSE, 
                      sample = min(rowSums(otu_tab), 
                                   col = "blue", cex = 0.6))
dev.off()

set.seed(90210)
ps_16S_rare <- rarefy_even_depth(ps_16S, sample.size = 15000)

alpha_16S <- alpha(ps_16S_rare, index = "all")
rownames(alpha_16S) <- sample_data(ps_16S)$name
alpha_16S$source <- sample_data(ps_16S)$soil
alpha_16S$name <- sample_data(ps_16S)$label
alpha_16S <- alpha_16S %>%
  relocate(source, name, .before = observed)

write.table(alpha_16S,
            file = "./02_out/alpha_diversity.csv",
            sep = ",",
            quote = T,
            row.names = T,
            col.names = T
            )
```
```{r}
ordcap = ordinate(ps_filter(ps_16S, GC != "NA"), "CAP", "bray", ~GC)
plot_ordination(ps_16S, ordcap, "samples", color="soil")
```

```{r shannon, fig.height=4, fig.width=12}
shannon_plot <- plot_richness(ps_merge,
                                  x = 'label',
                                  measures=c("Shannon"),
                                  title = "Shannon Diversity of Soil Community Fractions") +
  scale_x_discrete(labels = c(sample_data(ps_merge)$label)) +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x") +
  scale_color_npg() +
  guides(colour = "none") +
  labs(x = "",
       y = "Shannon Diversity") +
  theme(axis.text.x = element_text(angle = -90))
```


```{r observed, fig.height=4, fig.width=12}
observed_plot <- plot_richness(ps_merge,
                                  x = 'label',
                                  measures=c("Observed"),
                                  title = "Observed Richness of Soil Community Fractions") +
  scale_x_discrete(labels = c(sample_data(ps_merge)$label)) +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x") +
  scale_color_npg() +
  guides(colour = "none") +
  labs(x = "",
       y = "Richness") +
  theme(axis.text.x = element_text(angle = -90))
```

## Community Composition

When displaying the taxonomic makeup of the communities at phylum level, I split Pseudomonadota into classes.

```{r proteobacteria_phylum, warning=FALSE}
ps_merge_proteo <- ps_merge

plot.tax <- as.data.frame(tax_table(ps_merge))
plot.tax <-  data.frame(lapply(plot.tax, as.character), stringsAsFactors = F)
plot.tax$Phylum[plot.tax$Phylum=="Proteobacteria"] = plot.tax$Class[plot.tax$Phylum=="Proteobacteria"] 
plot.tax[] = lapply(plot.tax, factor)
plot.tax.table = tax_table(plot.tax)
rownames(plot.tax.table) = rownames(tax_table(ps_merge))
colnames(plot.tax.table) = colnames(tax_table(ps_merge))
tax_table(ps_merge_proteo) = plot.tax.table
rm(list = c("plot.tax" , "plot.tax.table"))

ps_16S_proteo <- ps_16S

plot.tax <- as.data.frame(tax_table(ps_16S))
plot.tax <-  data.frame(lapply(plot.tax, as.character), stringsAsFactors = F)
plot.tax$Phylum[plot.tax$Phylum=="Proteobacteria"] = plot.tax$Class[plot.tax$Phylum=="Proteobacteria"] 
plot.tax[] = lapply(plot.tax, factor)
plot.tax.table = tax_table(plot.tax)
rownames(plot.tax.table) = rownames(tax_table(ps_16S))
colnames(plot.tax.table) = colnames(tax_table(ps_16S))
tax_table(ps_16S_proteo) = plot.tax.table
rm(list = c("plot.tax" , "plot.tax.table"))

phylum_levels <- append(levels(factor(tax_table(ps_merge_proteo)[,"Phylum"])), "Other")
```

```{r gtdb_relabund}
write.table(ps_16S %>%
              transform_sample_counts(function(x) x / sum(x)) %>%
              tax_glom(taxrank = "Class", bad_empty = "Bacteria_unidentified") %>%
              psmelt() %>%
              select(Class, name, Abundance) %>%
              spread(name, Abundance),
            file = "./02_out/ps_16S.gtdb.relative_abundance.Class.csv",
            sep = ",",
            quote = F,
            row.names = F,
            col.names = T
            )
```

```{r phylum_relabund, fig.height=6, fig.width=20}
phylum_levels <- append(levels(factor(tax_table(ps_merge)[,"Phylum"])), "Other")
write.table(ps_merge %>%
              tax_glom(taxrank = "Phylum", bad_empty = "unidentified") %>%
              psmelt() %>%
              select(Phylum, Sample, Abundance) %>%
              spread(Sample, Abundance),
            file = "./02_out/ps_16S.relative_abundance.phylum.csv",
            sep = ",",
            quote = F,
            row.names = F,
            col.names = T
            )

ps <- ps_merge %>%
  transform_sample_counts(function(x) ((x * 100) / sum(x))) %>%
  tax_glom(taxrank = "Phylum", bad_empty = "unidentified") %>%
  aggregate_rare("Phylum", detection = 0.4, prevalence = 0.4)
#  transform_sample_counts(function(x) x * 100)

my_pal <- c(pal_simpsons("springfield")(16), pal_futurama("planetexpress")(4))

taxa_plot <- plot_bar(ps,
                          x = "label", 
                          fill="Phylum") +
  #scale_x_discrete(labels = c(sample_data(ps)$label)) +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x") +
  labs(fill = "Taxonomy (phylum level)",
       y = "Relative Abundance (%)") +
  theme() +
  scale_fill_manual(values = my_pal) +
  guides(fill = guide_legend(ncol = 1))
taxa_plot$data$Phylum <- factor(taxa_plot$data$Phylum, levels = phylum_levels)
taxa_plot$layers[[1]]$geom_params$width <- 1
```

Filler

## Plots for publication

Download links below plot

### Fig 1

```{r pub_fig_1, fig.height=6, fig.width=13}
zp1 <- shannon_plot +
  scale_y_continuous(limits = c(4.2, 6.8),
                     breaks = c(5, 6)) +
  labs(title = "Catchy Graph Title A",
       y = "Shannon diversity") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "italic"),
        strip.text = element_text(size = 12),
        plot.margin = unit(c(1,1,-0.1,1), "lines"))
zp2 <- taxa_plot +
  guides(fill = guide_legend(nrow = 3)) +
  labs(x = "",
       y = "Relative Abundance (%)",
       legend = "Phylum") +
  theme(strip.text.x = element_blank(),
        legend.position = "bottom",
        legend.spacing.x = unit(0.5, 'cm'),
        plot.margin = unit(c(-0.1,1,1,1), "lines"))
zp <- ggarrange(zp1, zp2,
          nrow = 2,
          heights = c(0.32,1),
          align = "v")
ggplot2::ggsave(zp, file = "./02_out/02_phyla_diversity.png", height = 7, width = 14)
```

```{r pub_fig_2, fig.height=6, fig.width=13}
zp1 <- observed_plot +
  scale_y_continuous(limits = c(100, 1400),
                     breaks = c(500, 1000)) +
  labs(title = "Catchy Graph Title A",
       y = "OTUs") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "italic"),
        strip.text = element_text(size = 12),
        plot.margin = unit(c(1,1,-0.1,1), "lines"))
zp2 <- taxa_plot +
  guides(fill = guide_legend(nrow = 2)) +
  labs(x = "Sample",
       y = "Relative Abundance",
       legend = "Phylum") +
  theme(strip.text.x = element_blank(),
        legend.position = "bottom",
        legend.spacing.x = unit(0.5, 'cm'),
        plot.margin = unit(c(-0.1,1,1,1), "lines"))
zp <- ggarrange(zp1, zp2,
          nrow = 2,
          heights = c(0.32,1),
          align = "v")
ggplot2::ggsave(zp, file = "./02_out/03_phyla_richness.png", height = 7, width = 14)
```
```{r}
ps <- ps_merge %>%
  transform_sample_counts(function(x) ((x * 100) / sum(x)))
my_subset <- subset(t(otu_table(ps)), rownames(t(otu_table(ps))) %in% mag_asv)
ps_mag <- merge_phyloseq(my_subset, tax_table(ps), sample_data(ps), refseq(ps_16S))
```


[Download this plot.](https://github.com/bodington/gc_16S/raw/main/02_out/01_alpha_diversity.png)
