---
title: "Soil GC Fractions Community Diversity"
author: "Dylan Bodington"
date: "`r Sys.Date()`"
out:
  html_document:
    toc: true
    toc_depth: 2
    code_folding: hide
    fig_caption: true
    
citation_package: bibtex
bibliography: bib.bib
link-citations: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(ggpubr)
library(ggsci)
library(RColorBrewer)
library(microshades)
library(tidyr)
library(purrr)
library(stringr)
library(dplyr)
library(phyloseq)
library(microbiome)
library(speedyseq)
library(microViz)
library(metagMisc)
library(PhyloMeasures)
library(Biostrings)
library(phyloseq.mate)
```

## Introduction

After cleaning and filtering using trimgalore, the 16S were processed using the standard [dada2 pipeline](https://benjjneb.github.io/dada2/tutorial.html) to create phyloseq objects of all ASVs. When rdp cannot reliably assign taxonomy (in this case down to Genus), levels below the lowest reliable assignment are designated NA.
[Download link for sequence files and metadata.](01_data/01_input)

```{r ps_objects, message=FALSE, warning=FALSE, include=FALSE}
load(file = "./01_data/phyloseq_all.Rdata")

## Labels for plots
labels_soil <- c("GcA (Craibstone pH 4.5)", "GcB (Craibstone pH 7.5)", "GcC (Corsemaul)", "GcL (Lower Dell)")
names(labels_soil) <- c("GcA", "GcB", "GcC", "GcL")

labels_fractioned <- c("0", "1")
names(labels_fractioned) <- c("Whole", "Fraction")
```

```{r merge_phyloseq,  message=FALSE, warning=FALSE, include=FALSE}
ps_raw <- readRDS('./01_data/phyloseq_gtdb.Rds')

tax_merge <- merge_phyloseq(tax_table(ps_raw),tax_table(as.matrix(read.table(file = "01_data/gc_mag_tax.csv", header = TRUE, sep = ",", row.names = 1))))
sample_data_merge <- sample_data(read.table(file = "00_ref/phyloseq_metadata.csv", header = TRUE, sep = ",", row.names = 1))

otu_mag <- otu_table(round(as.matrix(read.table(file = "01_data/gc_mag_otu.csv", header = TRUE, sep = ",", row.names = 1))), taxa_are_rows = FALSE)
otu_mag_merge_ps <- otu_table(matrix(0, length(rownames(otu_mag)), ncol(otu_table(ps_raw)), dimname = list(rownames(otu_mag),colnames(otu_table(ps_raw)))), taxa_are_rows = FALSE)
otu_ps_merge_mag <- otu_table(matrix(0, length(rownames(otu_table(ps_raw))), ncol(otu_mag), dimname = list(rownames(otu_table(ps_raw)),colnames(otu_mag))), taxa_are_rows = FALSE)
otu_merge <- merge_phyloseq(merge_phyloseq(otu_table(ps_raw), otu_mag_merge_ps), merge_phyloseq(otu_ps_merge_mag, otu_mag))

ps_merge <- phyloseq(otu_merge, sample_data_merge, tax_merge)

ps_set <- subset_samples(ps_merge, sample %in% c("4.5_M5", "4.5_M11", "7.5_M4", "7.5_M14", "CM_M5", "CM_M11", "LD_M4", "LD_M13"))

sample_names(ps_set) <- c("4.5_1", "4.5_20", "7.5_1", "7.5_20", "CM_1", "CM_20", "LD_1", "LD_20")

ps_merge <- merge_phyloseq(ps_merge, ps_set)

sample_data(ps_merge) <- sample_data(read.table(file = "00_ref/phyloseq_metadata.csv", header = TRUE, sep = ",", row.names = 1))


## Set factors for sample data and sort by name (for plot label sorting)
sample_data(ps_merge)$name <- factor(sample_data(ps_merge)$name, levels = str_sort(sample_data(ps_merge)$name, numeric = TRUE))
ps_merge <- ps_arrange(ps_merge, name)
sample_data(ps_merge)$label <- factor(sample_data(ps_merge)$label, levels = unique(sample_data(ps_merge)$label))
sample_data(ps_merge)$soil <- factor(sample_data(ps_merge)$soil)
sample_data(ps_merge)$amplicon <- factor(sample_data(ps_merge)$amplicon)
sample_data(ps_merge)$fractioned <- factor(sample_data(ps_merge)$fractioned)
sample_data(ps_merge)$fraction_no <- factor(sample_data(ps_merge)$fraction_no)



ps_16S <- subset_samples(ps_merge, amplicon == 1) %>%
  merge_phyloseq(refseq(ps_raw))

ps_merge <- subset_samples(ps_merge, !sample %in% c("4.5_0.1", "4.5_0.2", "4.5_0.3", "7.5_0.1", "7.5_0.2", "7.5_0.3", "CM_0.1", "CM_0.2", "CM_0.3", "LD_0.1", "LD_0.2", "LD_0.3"))

phyloseq_write_dataset(ps_16S, "./01_data/06_working/", "ps_16S")

## External packages to generate a 16S tree and calculate ASV novelty
#mafft --thread 8 ps_16S_ASVs.fasta > ps_16S_ASVs_aln.fasta
#trimal -in ps_16S_ASVs_aln.fasta -out ps_16S_ASVs_aln_nogaps.fasta -gappyout
#fasttree -nt -gtr -gamma ps_16S_ASVs_aln_nogaps.fasta > ps_16S_ASVs.tre
#

## Import ASV tree and merge into phyloseq object
asv_tree <- read_tree("01_data/06_working/ps_16S_ASV.tre")
ps_16S <- merge_phyloseq(ps_16S, asv_tree) %>%
  phyloseq_root_tree()

## Import ASV novelty as sample data
novel_asvs <- alpha(prune_taxa(colnames(otu_table(ps_16S)) %in% readLines("./01_data/06_working/ps_16S_ASV_novel.list"), ps_16S), index = "observed")/alpha(ps_16S, index = "observed")
sample_data(ps_16S) <- merge_phyloseq(sample_data(ps_16S), sample_data(novel_asvs %>% rename("observed" = "novel")))

save(ps_16S, ps_merge, file = "./01_data/phyloseq_all.Rdata")
```

## Alpha Diversity

```{r rarefaction_curve, fig.height=8, fig.width=12}
ps <- ps_16S

otu <- t(abundances(ps))
rare_curve <- vegan::rarecurve(otu, step = 50)
rare <- lapply(rare_curve, function(x){
  b <- as.data.frame(x)
  b <- data.frame(OTU = b[,1], raw.read = rownames(b))
  b$raw.read <- as.numeric(gsub("N", "",  b$raw.read))
  return(b)
  }
)
names(rare) <- rownames(otu)

rare <- map_dfr(
          rare,
          function(x){
            z <- data.frame(x)
            return(z)
            },
          .id = "sample"
          )
rare$soil = str_extract(rare$sample, pattern = "^[^_]+")

rare_labels <- as_labeller(c(`4.5` = "GcA", `7.5` = "GcB", `CM` = "GcC", `LD` = "GcL"))

rare_data <- rare %>%
            group_by(sample) %>%
            arrange(desc(raw.read)) %>%
            filter(row_number()==1)

rare_plot <- ggplot(rare) +
        geom_line(
          aes(
            x = raw.read,
            y = OTU,
            colour = sample
          )
        ) +
        geom_text(
          data = rare_data,
          aes(
            x = raw.read + 6000, y = OTU, label = rare_data$sample
          )
        ) +
        facet_wrap(
          vars(soil),
          nrow = 1,
          labeller = rare_labels
        )
```

```{r novelty}

```

```{r alpha_16S, fig.height=8, fig.width=12}
ps <- ps_16S

alpha_all <- alpha(ps, index = "all")
alpha_all$diversity_faith <- phyloseq_phylo_div(ps, measures = c("PD"))[,1]
alpha_all$asv_novelty <- sample_data(ps_16S)$novel
rownames(alpha_all) <- sample_data(ps)$name
alpha_all$source <- sample_data(ps)$soil
alpha_all$name <- sample_data(ps)$label
alpha_all <- alpha_all %>%
  relocate(source, name, .before = observed)

write.table(alpha_all,
            file = "./02_out/alpha_diversity.csv",
            sep = ",",
            quote = T,
            row.names = T,
            col.names = T
            )
```

```{r shannon, fig.height=4, fig.width=12}
ps <- subset_samples(ps_merge, !sample %in% c("4.5_20", "4.5_M5", "4.5_M11", "7.5_20", "7.5_M4", "7.5_M14", "CM_20", "CM_M5", "CM_M11", "LD_20", "LD_M4", "LD_M13"))

shannon_plot <- plot_richness(ps,
                                  x = 'label',
                                  measures=c("Shannon")
                                  ) +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x",
             labeller = labeller(soil = labels_soil)
             ) +
  scale_color_npg() +
  guides(colour = "none") +
  labs(x = "",
       y = "Shannon\ndiversity") +
  theme_bw(
    axis.text.x = element_text(angle = -90)
    )
```


```{r observed, fig.height=4, fig.width=12}
ps <- subset_samples(ps_merge, !sample %in% c("4.5_20", "4.5_M5", "4.5_M11", "7.5_20", "7.5_M4", "7.5_M14", "CM_20", "CM_M5", "CM_M11", "LD_20", "LD_M4", "LD_M13"))

observed_plot <- plot_richness(ps,
                                  x = 'label',
                                  measures=c("Observed"),
                                  ) +
  facet_wrap(~soil,
             ncol = 4,
             scales="free_x",
             labeller = labeller(soil = labels_soil)
             ) +
  scale_color_npg() +
  guides(colour = "none") +
  labs(x = "",
       y = "Richness") +
  theme_bw(
    axis.text.x = element_text(angle = -90)
    )
```

```{r unifrac}
ps <- subset_samples(ps_16S, !sample %in% c("4.5_0.1", "4.5_0.2", "4.5_0.3", "7.5_0.1", "7.5_0.2", "7.5_0.3", "CM_0.1", "CM_0.2", "CM_0.3", "LD_0.1", "LD_0.2", "LD_0.3"))

unifrac = ordinate(ps, "PCoA", "unifrac", ~GC)
unifrac_plot <- plot_ordination(ps,
                  unifrac,
                  "samples",
                  color = "soil",
                  shape = "fractioned"
                  ) +
                  theme_bw() +
                  scale_color_npg() +
                  labs(
                    color = "Soil",
                    shape = "Fraction"
                  ) +
                  scale_shape_manual(
                    name = "Fraction",
                    labels = names(labels_fractioned),
                    values = c(23, 16)
                  ) +
                  geom_point(
                    size = 4
                  )

ggplot2::ggsave(unifrac_plot, file = "./02_out/test_unifrac.png", height = 5, width = 6)
```

## Community Composition

```{r gtdb_relabund}
#phylum_levels <- append(sort_taxa$Phylum[nzchar(sort_taxa$Phylum)], c("Unassigned", "Other"))
ps <- ps_merge

write.table(ps %>%
              tax_fix() %>%
              tax_transform(trans = "compositional", rank = "Phylum") %>%
              psmelt() %>%
              select(name, Phylum, Abundance) %>%
#              arrange(relabund_table, match(relabund_table$Phylum, phylum_levels)) %>%
              pivot_wider(names_from = Phylum, values_from = Abundance),
            file = "./02_out/ps_16S.relative_abundance.phylum.csv",
            sep = ",",
            quote = F,
            row.names = F,
            col.names = T
            )
```

```{r phylum_relabund, fig.height=6, fig.width=20}
ps <- ps_merge %>%
  filter_taxa2(~ sum(. > 0) > 0) %>%
  mutate_tax_table(across(everything(), \(x) str_replace(x,'_[A-Z]', ''))) %>%
  subset_taxa(Kingdom != "")

phylum_levels <- append(levels(factor(tax_table(ps)[,"Phylum"])), c("Other"))

ps <- ps %>%
  tax_fix() %>%
  tax_transform(trans = "compositional") %>%
  transform_sample_counts(function(x) x * 100) %>%
  tax_agg(rank = "Phylum", top_N = 24)

ps_frac <- subset_samples(ps, !sample %in% c("4.5_20", "4.5_M5", "4.5_M11", "7.5_20", "7.5_M4", "7.5_M14", "CM_20", "CM_M5", "CM_M11", "LD_20", "LD_M4", "LD_M13"))
#ps_frac <- subset_samples(ps, !sample %in% c("4.5_1", "4.5_20", "4.5_M5", "4.5_M11", "7.5_1", "7.5_20", "7.5_M4", "7.5_M14", "CM_1", "CM_20", "CM_M5", "CM_M11", "LD_1", "LD_20", "LD_M4", "LD_M13"))

ps_mag <- subset_samples(ps, sample %in% c("4.5_M5", "4.5_M11", "7.5_M4", "7.5_M14", "CM_M5", "CM_M11", "LD_M4", "LD_M13"))

relabund_pal <- c(pal_simpsons("springfield")(16)[-5], "steelblue2", pal_flatui("default")(8)[-5], "dimgray")
```

```{r frac_relabund, fig.height=6, fig.width=20}
frac_taxa_plot <- plot_bar(ps_frac,
                          x = "label", 
                          fill="top") +
  facet_wrap(
    ~soil,
    ncol = 4,
    scales="free_x",
    labeller = labeller(soil = labels_soil)
    ) +
  labs(
    fill = "Taxonomy (Phylum)",
    y = "Relative Abundance (%)"
    ) +
  scale_fill_manual(
    values = relabund_pal
    ) +
  guides(
    fill = guide_legend(
      ncol = 1,
      byrow = TRUE)
    ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = -90)
    )

frac_taxa_plot$data$top[frac_taxa_plot$data$top %in% c("Bacteria Kingdom", "other", NA)] <- rep("Other", length(frac_taxa_plot$data$top[frac_taxa_plot$data$top %in% c("Bacteria Kingdom", "other", NA)]))
frac_taxa_plot$data$top <- str_pad(frac_taxa_plot$data$top, 21, "right")
phylum_levels <- str_pad(phylum_levels, 21, "right")
frac_taxa_plot$data$top <- factor(frac_taxa_plot$data$top, levels = phylum_levels)
frac_taxa_plot$layers[[1]]$geom_params$width <- 1

ggsave(frac_taxa_plot, file = "./02_out/test_frac_relabund.png", height = 8, width = 16)
```

```{r mag_relabund, fig.height=6, fig.width=20}
mag_taxa_plot <- plot_bar(ps_mag,
                          x = "label", 
                          fill="top") +
  facet_wrap(
    ~soil,
    ncol = 4,
    scales="free_x"
    ) +
  labs(
    fill = "Taxonomy (Phylum)    ",
    y = "Relative Abundance (%)"
    ) +
  scale_fill_manual(
    values = relabund_pal
    ) +
  guides(
    fill = guide_legend(
      ncol = 1,
      byrow = TRUE
      )
    ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = -90)
    )

mag_taxa_plot$data$top[mag_taxa_plot$data$top %in% c("Bacteria Kingdom", "other", NA)] <- rep("Other", length(mag_taxa_plot$data$top[mag_taxa_plot$data$top %in% c("Bacteria Kingdom", "other", NA)]))
mag_taxa_plot$data$top <- str_pad(mag_taxa_plot$data$top, 21, "right")
phylum_levels <- str_pad(phylum_levels, 21, "right")
mag_taxa_plot$data$top <- factor(mag_taxa_plot$data$top, levels = phylum_levels)
mag_taxa_plot$layers[[1]]$geom_params$width <- 1

ggsave(mag_taxa_plot, file = "./02_out/test_mag_relabund.png", height = 8, width = 8)
```

Filler

## Plots for publication

Download links below plot

### Fig 1

```{r pub_fig_1, fig.height=6, fig.width=13}
ap1 <- shannon_plot +
  scale_y_continuous(limits = c(4.2, 6.8),
                     breaks = c(5, 6)) +
  labs(title = "a") +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_text(size = 12),
    plot.margin = unit(c(1,1,-0.1,1), "lines"))

ap2 <- observed_plot +
  scale_y_continuous(limits = c(100, 1300),
                     breaks = c(500, 1000)) +
  labs(y = "Richness") +
  theme(
    title = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_blank(),
    plot.margin = unit(c(-0.1,1,-0.1,1), "lines"))

ap3 <- frac_taxa_plot +
  guides(
    fill = guide_legend(
      nrow = 3,
      byrow = TRUE
      )
    ) +
  labs(
    x = "",
    y = "Relative Abundance (%)"
    ) +
  theme(
    strip.text.x = element_blank(),
    axis.text.x = element_text(
      size = 8,
      colour = "black",
      vjust = 0.5
      ),
    legend.position = "bottom",
    legend.title = element_text(size = 14),
    legend.title.position = "top",
    legend.title.align = 0.5,
    legend.justification = "right",
    legend.spacing.x = unit(0.5, 'cm'),
    legend.margin = margin(0.1,1,1,1),
    legend.text = element_text(size = 10),
    plot.margin = unit(c(-0.1,1,1,1), "lines")
    )

ap <- ggarrange(ap1, ap2, ap3,
          nrow = 3,
          heights = c(0.32, 0.22, 1),
          align = "v")

bp1 <- unifrac_plot +
  labs(title = "b") +
  theme(
    plot.margin = margin(
      l = 0.5,
      b = 0.5,
      unit = "cm"
    )
  )

bp2 <- mag_taxa_plot +
  labs(title = "c",
        x = "",
        y = "Relative Abundance (%)"
  ) +
  theme(axis.text.x = element_text(
          colour = "black",
          vjust = 0.5
        ),
        legend.position = "none",
        plot.margin = unit(c(-0.1,1,1,1), "lines")
    )  

bp <- ggarrange(bp1, NULL, bp2,
          ncol = 3,
          widths = c(0.72, 0.1, 0.6),
          align = "h")

zp <- ggarrange(ap, bp,
          nrow = 2,
          heights = c(1, 0.8),
          align = "v")

ggplot2::ggsave(zp, file = "./02_out/01_diversity.png", height = 14, width = 14)
```

[Download this plot.](https://github.com/bodington/gc_16S/raw/main/02_out/01_alpha_diversity.png)

```{r pub_fig_2, fig.height=6, fig.width=13}
zp <- rare_plot +
  labs(title = "Catchy Rare Graph Title",
        x = "Sample Size",
        y = "Observed OTUs") +
  scale_x_continuous(limits = c(0, 75000)) +
  theme(strip.text.x = element_text(size = 12),
        legend.position = "none",
        plot.margin = unit(c(1,1,1,1), "lines"))
ggplot2::ggsave(zp, file = "./02_out/02_rare.png", height = 7, width = 14)
```

[Download this plot.](https://github.com/bodington/gc_16S/raw/main/02_out/02_rare.png)

```{r pub_fig_s1, fig.height=5, fig.width=13}
zp1 <- shannon_plot +
  scale_y_continuous(limits = c(4.2, 6.8),
                     breaks = c(5, 6)) +
  labs(title = "Catchy Graph Title A",
       y = "Shannon\ndiversity") +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_text(size = 12),
    plot.margin = unit(c(1,1,-0.1,1), "lines"))

zp2 <- observed_plot +
  scale_y_continuous(limits = c(100, 1200),
                     breaks = c(500, 1000)) +
  labs(y = "Richness") +
  theme(
    title = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 10),
    strip.text = element_blank(),
    plot.margin = unit(c(1,1,-0.1,1), "lines"))

zp3 <- taxa_plot_domain +
  guides(fill = guide_legend(nrow = 1, byrow = TRUE)) +
  labs(x = "",
       y = "Relative Abundance (%)",
       legend = "Phylum") +
  theme(strip.text.x = element_blank(),
        axis.text.x = element_text(
          colour = "black",
          vjust = 0.5
          ),
        legend.position = "bottom",
        legend.justification = "right",
        legend.spacing.x = unit(0.5, 'cm'),
        legend.text = element_text(
          size = 9,
          margin = margin(l = 0.1, unit ="cm")
          ),
        plot.margin = unit(c(-0.1,1,1,1), "lines"))

zp <- ggarrange(zp1, zp2, zp3,
          nrow = 3,
          heights = c(0.32, 0.22, 0.6),
          align = "v")

ggplot2::ggsave(zp, file = "./02_out/s1_alpha_diversity.png", height = 8, width = 14)
```

```{r, message=FALSE, warning=FALSE, include=FALSE}
ps <- ps_merge %>%
  transform_sample_counts(function(x) ((x * 100) / sum(x)))
my_subset <- subset(t(otu_table(ps)), rownames(t(otu_table(ps))) %in% mag_asv)
ps_mag <- merge_phyloseq(my_subset, tax_table(ps), sample_data(ps), refseq(ps_16S))
```



